pkg/image/prune/prune.go

```diff
@@ -1190,3 +1160,62 @@ func getName(obj runtime.Object) string {
 	}
 	return fmt.Sprintf("%s/%s", accessor.GetNamespace(), accessor.GetName())
 }
+
+// tryProtocolsWithRegistryURL runs given callback with different protocols until no error is returned. The
+// https protocol is the first attempt. If it fails and allowInsecure is true, http will be the next. Obtained
+// errors will be concatenated and returned.
+func tryProtocolsWithRegistryURL(registry string, allowInsecure bool, callback func(registryURL url.URL) error) (*url.URL, error) {
+	var errs []error
+
+	if !strings.Contains(registry, "://") {
+		registry = "unset://" + registry
+	}
+	url, err := url.Parse(registry)
+	if err != nil {
+		return nil, err
+	}
+	var protos []string
+	switch {
+	case len(url.Scheme) > 0 && url.Scheme != "unset":
+		protos = []string{url.Scheme}
+	case allowInsecure || netutils.IsPrivateAddress(registry):
+		protos = []string{"https", "http"}
+	default:
+		protos = []string{"https"}
+	}
+	registry = url.Host
+	glog.V(4).Infof("url: %#+v", url)
+
+	for i, proto := range protos {
+		switch i {
+		case 0:
+			glog.V(4).Infof("Trying protocol %s for the registry url %s", proto, registry)
+		default:
+			glog.V(4).Infof("Falling back to protocol %s for the registry url %s", proto, registry)
+		}
```

Comment:
This is bad, just log the fact, you'll figure out when looking at logs, that this is a retry. There's no point in hardcoding this switch.